README

.. -*- restructuredtext -*-

################################################
Dragonkeeper - Opera Dragonfly development proxy
################################################

Introduction
------------

Dragonkeeper is an utility for working on the Opera Dragonfly client.
Dragonfly serves as a proxy between the debugger and the debuggee. This makes
it possible to run the Opera Dragonfly client as a normal web page. This
makes development easy, as on can simple reload the client whenever making
changes to it.

Howto
-----

Basic workflow when using Dragonkeeper is as follows:

- Get a source distribution of Opera Dragonfly from http://bitbucket.org/scope/dragonfly-stp-1/
- Launch Dragonkeeper with the appropriate options. If the current directory
  is the Dragonfly directory, all the default options should be fine.
- Open a browser to be debugged, the debuggee.
- In the debugee, open the adress "opera:debug".
- Enter the IP address and port of the machine running Dragonkeeper.
  The defaults are localhost and 7001.
- Open a separate Opera instance, this will be the debugger instance.
- Navigate to the IP/port of the Dragonkeeper instance. The default port is
  8002
- You'll get a directory listing. You should open client-en.xml in the src
  folder.
- This will load Opera Dragonfly from the Dragonkeeper and connect to the
  debuggee.
  
Note that both the debuggee and the debugger can run on the same machine. This
makes sense when developing.


Usage
-----

Usage: -m dragonkeeper.dragonkeeper [options]

Exit: Control-C

Settings:  an optional file CONFIG does overwrite the defaults.
The options file is a standard .ini file, with a single section called
"dragonkeeper":

[dragonkeeper]
host:
root: .
server_port: 8002
proxy_port: 7001
debug: False
format: False

Options:
  -h, --help            show this help message and exit
  -c CONFIG_PATH, --config=CONFIG_PATH
                        Path to config file
  -d, --debug           print message flow
  -f, --format          pretty print message flow
  -j, --format-payload  pretty print the message payload. can be very
                        expensive
  -r ROOT, --root=ROOT  the root directory of the server; default .
  -p PROXY_PORT, --proxy-port=PROXY_PORT
                        proxy port; default 7001
  -s SERVER_PORT, --server-port=SERVER_PORT
                        server port; default 8002
  --host=HOST           host; default localhost
  -i, --make-ini        Print a default dragonkeeper.ini and exit
  --force-stp-0         force stp 0 protocol
  --print-command-map   print the command map
  --message-filter=MESSAGE_FILTER
                        Filter the printing of the messages. The argument is
                        the filter or a path to a file with the filter. If the
                        filter is set, only messages which are listed in the
                        filter will be printed. The filter uses JSON notation
                        like: {"<service name>": {"<message type>":
                        [<message>*]}}", with message type one of "command",
                        "response", "event."  '*' placeholder are accepted in
                        <message>, e.g. a filter to log all threads may look
                        like:  "{'ecmascript-debugger': {'event':
                        ['OnThread*']}}".
  -v, --verbose         print verbose debug info
  --cgi                 enable cgi support


More comments in the source files.


Changelog
---------

See the CHANGELOG file

Contact
-------

Dragonkeeper is maintained by the Opera Dragonfly team. The authors are

- Christian Krebs <chrisk@opera.com>
- Rune Halvorsen <runeh@opera.com>
- Jan Borsodi <jborsodi@opera.com>

The Opera Dragonfly web site is at http://dragonfly.opera.com


License
-------

See the ``LICENSE`` file in the top distribution directory.





CHANGELOG

0.8
---

* Initial public release.
* (No changelog remains from internal development, but hg log is intact)

0.8.2
-----

* added support for cgi


proxy.py
import ConfigParser
from Connection import ScopeInterface
from ScopeConnection import ScopeConnection
from SimpleServer import SimpleServer, asyncore
from common import Options

CONFIG_FILE = "dragonkeeper.ini"

APP_DEFAULTS = {
    "host": "",
    "server_port": 8002,
    "proxy_port": 7001,
    "root": '.',
    "debug": False,
    "format": False
}

DEFAULT_TYPES = {
    "host": str,
    "server_port": int,
    "proxy_port": int,
    "root": str,
    "debug": bool,
    "format": bool
}

USAGE = """%prog [options]
    
Exit: Control-C
    
Settings:  an optional file CONFIG does overwrite the defaults.
The options file is a standard .ini file, with a single section called
"dragonkeeper":

[dragonkeeper]
host:
root: .
server_port: 8002
proxy_port: 7001
debug: False
format: False"""


def run_proxy(options, count=None): 
    SimpleServer(options.host, options.server_port, ScopeInterface, options)
    SimpleServer(options.host, options.proxy_port, ScopeConnection, options)
    print "server on: http://%s:%s/" % ( 
                options.host or "localhost", options.server_port )
    asyncore.loop(timeout = 0.1, count = count)

def _load_config(path):
    """Load an .ini file containing dragonkeeper options. Returns a dict
    with options. If file does not exist, or can't be parsed, return None """
    config = ConfigParser.RawConfigParser()

    okfile = config.read(path)
    if not okfile or not config.has_section("dragonkeeper"):
        return None

    ret = {}
    for name, value in config.items("dragonkeeper"):
        ret[name] = DEFAULT_TYPES[name](value)
            
    return ret

def _print_config():
    print "[dragonkeeper]"
    for item in APP_DEFAULTS.items():
        print "%s: %s" % item   

def _parse_options():
    """Parse command line options.
    
    The option priority is like so:
    Least important, app defaults
    More important, settings in config file
    Most important, command line options
    """
    from optparse import OptionParser

    parser = OptionParser(USAGE)
    parser.add_option(
        "-c", "--config",
        dest = "config_path", 
        help = "Path to config file"
    )
    parser.add_option(
        "-d", "--debug",
        action = "store_true", 
        dest = "debug", 
        help = "print message flow"
        )
    parser.add_option(
        "-f", "--format",
        action="store_true", 
        dest = "format", 
        help = "pretty print message flow"
        )
    parser.add_option(
        "-r", "--root", 
        dest = "root", 
        help = "the root directory of the server; default %s" % (
                    APP_DEFAULTS["root"])
        )
    parser.add_option(
        "-p", "--proxy-port",
        dest = "proxy_port", 
        type="int",
        help = "proxy port; default %s" % APP_DEFAULTS["proxy_port"]
        )
    parser.add_option(
        "-s", "--server-port",
        dest = "server_port", 
        type="int",
        help = "server port; default %s" % APP_DEFAULTS["server_port"]
        )
    parser.add_option(
        "--host",
        dest = "host", 
        help = "host; default %s" % APP_DEFAULTS["host"]
        )
    parser.add_option(
        "-i", "--make-ini",
        dest = "make_ini", 
        action="store_true",
        default=False,
        help = "Print a default dragonkeeper.ini and exit"
        )
    options, args = parser.parse_args()

    if options.make_ini:
        _print_config()
        sys.exit(0)

    # appopts contains the defaults
    appopts = Options(APP_DEFAULTS)

    if options.config_path: # an explicit config file was given. Overrides everything
        config = _load_config(options.config_path)
        if config:
            appopts = Options(config)
        else:
            parser.error("""Invalid path or config file "%s"!""" % options.config_path)

    elif os.path.isfile(CONFIG_FILE): # if not explicit config, try to load default ini file.
        config = _load_config(CONFIG_FILE)
        if config:
            for key, value in config.items():
                appopts[key] = value

    # at this point we have an appopts object with all we need. A mix 
    # of defaults and the .ini files
    # Any command line options will override this.

    for name, value in [(e, getattr(options, e)) for e in APP_DEFAULTS.keys()]:
        if not value is None:
            appopts[name] = value

    # All options set. Now we can check if they are OK
    
    if not os.path.isdir(appopts.root):
        parser.error("""Root directory "%s" does not exist""" % options.root)
        
    return appopts

if __name__ == "__main__":
    import sys
    import os

    options = _parse_options()

    os.chdir(options.root) 
    try:
        run_proxy(options)
    except KeyboardInterrupt:
        # todo: shut down the open connections cleanly
        pass
    """
    import cProfile, sys
    p=open("profile", "w")
    sys.stdout = p
    cProfile.run("run_proxy(count = 5000, context = options)")
    p.close()
    """
  
